[exporter/datasetexporter]: Make export of resources and scopes more flexible

martin-majlis-s1 opened the pull request:
**Description:** Make export of resources and scopes more flexible

**Link to tracking Issue:** #27651 , #27649 

**Testing:** Unit tests

**Documentation:** <Describe the documentation added.>

**Note:**: This PR is on top of this PR - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27663/files - since it's introducing some helper functions and it's fixing NPE.

Fixes #27651
Fixes #27649

martin-majlis-s1 committed:
[exporter/datasetexporter]: Fix NPE exception when attribute contains nil

martin-majlis-s1 committed:
Merge branch 'main' into 27648-dataset-fix-npe

martin-majlis-s1 committed:
[exporter/datasetexporter]: Make export of resources and scopes more flexible

martin-majlis-s1 committed:
Add Changelog entry

martin-majlis-s1 committed:
Merge branch 'main' into 27648-dataset-fix-npe

martin-majlis-s1 committed:
Merge branch '27648-dataset-fix-npe' into 27651-make-resource-scope-more-flexible

martin-majlis-s1 committed:
Update documentation

martin-majlis-s1 commented:
This PR is built on top of #27663 since that PR is fixing NPE and this PR is using the helper function as well it tests.

martin-majlis-s1 committed:
Merge branch 'main' into 27651-make-resource-scope-more-flexible

martin-majlis-s1 committed:
Add configuration options for suffixes and complex bodies

martin-majlis-s1 committed:
Merge branch 'main' into 27651-make-resource-scope-more-flexible

martin-majlis-s1 commented:
Majority of the changed lines are in the tests since I have introduced few configuration options and I have to test various combinations.

peteryork commented on exporter/datasetexporter/README.md:
```diff
@@ -43,6 +43,17 @@ The process of populating the serverHost attribute works as follows:
 
+### Attributes
+
+If exporting of attributes of given type are enabled, then they are exported in following order:
```
Enabled attributes are exported in the order:

martin-majlis-s1 replied:
Fixed

peteryork commented on exporter/datasetexporter/README.md:
```diff
@@ -43,6 +43,17 @@ The process of populating the serverHost attribute works as follows:
+3. Scope attributes
+4. Log attributes
+
+In case of name conflict the later attribute name will be appended with the `export_distinguishing_suffix` value. If the value is empty string then the last value wins.
```
If there is a name conflict, the `export_distinguishing_suffix` value is appended to the later attribute's name. If the the `export_distinguishing_suffix` value is an empty string, then the preceding attribute's name is used. 

I think you mean preceding here - the name of the attribute in the hierarchy before the attribute with the naming conflict and empty `export_distinguishing_suffix`?

martin-majlis-s1 replied:
It's indeed confusing. I will add some examples :)

> If there is a name conflict, the export_distinguishing_suffix value is appended to the later attribute's name. If the the export_distinguishing_suffix value is an empty string, then the value from the last attribute is used.

martin-majlis-s1 replied:
I have added this as separate line in the attributes section.

peteryork commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
+    - `export_scope_info_on_event` (default = true): Include LogRecord scope information (if available) on the DataSet event.
+    - `export_scope_prefix` (default = 'scope.attributes.'): If scope is included then the following string will be used as prefix.
+    - `export_separator` (default = '.'): When attribute contains some nested structures (maps, arrays) this string is used to separate them during flattening.
+    - `export_distinguishing_suffix` (default = '_'): If there is a collision between event attributes the following string is used as suffix for distinguishing them.
```
A suffix string to resolve naming collisions when flattening with `export_separator`.

martin-majlis-s1 replied:
Fixed

martin-majlis-s1 committed:
Merge branch 'main' into 27651-make-resource-scope-more-flexible

martin-majlis-s1 commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
   - `retry_shutdown_timeout` (default = 30s): The maximum time for which it will try to send data to the DataSet during shutdown. This value should be shorter than container's grace period.
 - `logs`:
-    - `export_scope_info_on_event` (default = false): Include LogRecord scope information (if available) on the DataSet event.
+    - `export_resource_info_on_event` (default = false): Include LogRecord resource information (if available) on the DataSet event.
```
From those attributes we are constructing the DataSet event. I will keep it as is. When I add examples, it will be maybe clearer, and then we can adjust the wording.

martin-majlis-s1 commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
 - `logs`:
-    - `export_scope_info_on_event` (default = false): Include LogRecord scope information (if available) on the DataSet event.
+    - `export_resource_info_on_event` (default = false): Include LogRecord resource information (if available) on the DataSet event.
+    - `export_resource_prefix` (default = 'resource.attributes.'): If resource is included then the following string will be used as prefix.
```
Fixed. :)

martin-majlis-s1 commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
-    - `export_scope_info_on_event` (default = false): Include LogRecord scope information (if available) on the DataSet event.
+    - `export_resource_info_on_event` (default = false): Include LogRecord resource information (if available) on the DataSet event.
+    - `export_resource_prefix` (default = 'resource.attributes.'): If resource is included then the following string will be used as prefix.
+    - `export_scope_info_on_event` (default = true): Include LogRecord scope information (if available) on the DataSet event.
```
Same as above.

martin-majlis-s1 commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
+    - `export_resource_info_on_event` (default = false): Include LogRecord resource information (if available) on the DataSet event.
+    - `export_resource_prefix` (default = 'resource.attributes.'): If resource is included then the following string will be used as prefix.
+    - `export_scope_info_on_event` (default = true): Include LogRecord scope information (if available) on the DataSet event.
+    - `export_scope_prefix` (default = 'scope.attributes.'): If scope is included then the following string will be used as prefix.
```
Fixed. :)

martin-majlis-s1 commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
     - `decompose_complex_message_field` (default = true): Set this to false to disable decomposing complex body / message field types (e.g. a map) into separate fields.
+    - `decomposed_complex_message_prefix` (default = 'body.map.'): If complex message field is decomposed then the following string will be used as prefix.
+- `traces`:
+    - `export_separator` (default = '.'): When attribute contains some nested structures (maps, arrays) this string is used to separate them during flattening.
```
Fixed

martin-majlis-s1 commented on exporter/datasetexporter/README.md:
```diff
@@ -53,8 +64,17 @@ Make sure to provide the appropriate server host value in the `serverHost` attri
+    - `decomposed_complex_message_prefix` (default = 'body.map.'): If complex message field is decomposed then the following string will be used as prefix.
+- `traces`:
+    - `export_separator` (default = '.'): When attribute contains some nested structures (maps, arrays) this string is used to separate them during flattening.
+    - `export_distinguishing_suffix` (default = '_'): If there is a collision between event attributes the following string is used as suffix for distinguishing them.
```
Fixed.

martin-majlis-s1 committed:
Add examples to the documentation to make it more clear.

peteryork reviewed (approved):
LGTM. Thanks.

martin-majlis-s1 committed:
Merge branch 'main' into 27651-make-resource-scope-more-flexible

mx-psi commented on exporter/datasetexporter/config.go:
```diff
@@ -17,41 +17,84 @@ import (
+const tracesExportDistinguishingSuffix = "_"
+
 type TracesSettings struct {
+	// ExportSeparator is separator used when flattening exported attributes
+	// Default value: .
+	ExportSeparator string `mapstructure:"export_separator"`
+	// ExportDistinguishingSuffix is suffix used to be appended to the end of attribute name in case of collision
+	// Default value: _
+	ExportDistinguishingSuffix string `mapstructure:"export_distinguishing_suffix"`
```
nit: move these to a separate struct that you embed on `TraceSettings` and `LogsSettings` to avoid duplication both on the struct definition and the default definition

martin-majlis-s1 replied:
This was my initial thinking as well, that this should be single configuration shared by logs and traces. 

But then I have realised, that:
1.  the logic for logs and traces is completely different. For traces we are extracting just handful of properties from the resource and we completely ignore the scope. For logs people can extract everything.
2. Most people are extracting either logs or traces. So it's nice if you can define it in "your" section.
3. It was not obvious to me, what should be the name of the shared section - `export`, `extraction` - looks little bit strange to me.
4. Initially I was thinking, that none of these two things should be configurable. But if you are specifying prefixes, maybe you want to also specify the separator -> so I have made it configurable.

I do not think that any of those solution is significantly better. If it's fine for you, can we keep it as is - and we can observe, how people use it.

mx-psi replied:
I think it's fine if they are separate on the configuration exposed to the user because of the points you mentioned, I tend to be conservative when adding new options, but this is a personal preference and I don't want to impose that on you. It is also unclear to me what the name of the shared section should be if we make it a common section.

What I do think we can make is to avoid duplication at the Go level. We can define a struct:

```go
type exportSettings struct {
	// ExportSeparator is separator used when flattening exported attributes
	// Default value: .
	ExportSeparator string `mapstructure:"export_separator"`
	// ExportDistinguishingSuffix is suffix used to be appended to the end of attribute name in case of collision
	// Default value: _
	ExportDistinguishingSuffix string `mapstructure:"export_distinguishing_suffix"`
}
```

which we can then embed in the `LogsSettings` and `TraceSettings`:

```go
type TracesSettings struct {
    exportSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
    // existing fields go here
}


type LogsSettings struct {
    exportSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
    // existing fields go here
}
```

and have a common helper for defining the default:

```go
func newDefaultExportSettings() exportSettings {
    return exportSettings{
          ExportSeparator:            exportSeparatorDefault,
	  ExportDistinguishingSuffix: ExportDistinguishingSuffix,
     }
}
```

which we can use for the default settings:
```go
func newDefaultLogsSettings() LogsSettings {
	return LogsSettings{
		ExportResourceInfo:             logsExportResourceInfoDefault,
		ExportResourcePrefix:           logsExportResourcePrefixDefault,
		ExportScopeInfo:                logsExportScopeInfoDefault,
		ExportScopePrefix:              logsExportScopePrefixDefault,
		exportSettings: newDefaultExportSettings(),
		DecomposeComplexMessageField:   logsDecomposeComplexMessageFieldDefault,
		DecomposedComplexMessagePrefix: logsDecomposedComplexMessageFieldPrefixDefault,
	}
}
```

Does that make sense?

martin-majlis-s1 replied:
Thanks! I haven't realised, that it's possible to do it like this. :)

tpaschalis committed:
Update Grafana Agent distribution components (#27615)

**Description:** This adds Grafana Agent as a distribution to some new
components after the Grafana Agent v0.37 release.

https://github.com/grafana/agent/blob/main/CHANGELOG.md#v0370-2023-10-10

**Link to tracking Issue:**  None filed.

**Testing:** None added.

**Documentation:** Adding links to grafana agent in the
alphabetically-sorted list.

---------

Signed-off-by: Paschalis Tsilias <paschalis.tsilias@gmail.com>
Co-authored-by: Juraci Paixão Kröhling <juraci@kroehling.de>

dmitryax committed:
[cmd/otecontribcol] Add sending data step to exporter lifecycle tests (#27824)

As the next step, I'll update it to send readonly data to immutable
exporters

andrzej-stencel committed:
[chore][exporter/syslog] docs: describe default values, add examples (#27831)

Fixes
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21242

dmitryax committed:
[chore] Send readonly data to immutable exporters in lifecycle tests (#27825)

This should help to catch exporters that are incorrectly claimed as not
mutating.

hovavza committed:
UDP input operator - async mode - separate between readers & processors (#27805)

**Description:** adding a feature - when async mode is enabled in the
UDP receiver (udp input operator), separating reading from processing
operations.
This is important to reduce data-loss in high scale UDP scenarios. See
original issue for more details.
The async config block is changed now. Instead of readers field
(determining the concurrency level of how many threads the udp receiver
is running, all reading from the UDP port, processing, and sending
downstream), it will now have 2 fields:
- readers - determines the concurrency level of threads only reading
from UDP port and pushing the packets to a channel.
- processors - determines the concurrency level of threads reading from
the channel, processing the packets, and sending downstream.
- max_queue_length - determines the max size of the channel between the
readers & the processors. Setting it high enough, allows to prevent
data-loss in cases of downstream temporary latency. Once channel is
full, the readers thread will stop until there's room in the queue (so
to prevent unlimited memory usage).
This improves performance and reduces UDP packet loss in high-scale
scenarios.
Note that async mode only supports this separation of readers from
processors. If async config block isn't included, the default state

**Link to tracking Issue:** 27613

**Testing:** Local stress tests ran all types of async config (no
'async', with 'async', etc.). Updating existing udp test accordingly.
Also, ran scale tests and saw improvement in data-loss.

**Documentation:** Updated md file for both udplogreceiver & stanza
udp_input operator with the new flags.

---------

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

BinaryFissionGames committed:
[receiver/filelog] Implement specifying top n files to track when ordering (#27844)

**Description:**
* Add a new `ordering_criteria.top_n` option, which allows a user to
specify the number of files to track after ordering.
  * Default is 1, which was the existing behavior.

**Link to tracking Issue:** #23788

**Testing:**
Unit tests added.

**Documentation:**
Added new parameter to existing documentation.

mdonkers committed:
[pkg/pdatautil] Optimize the hashing function for `pcommon.Map` (#27840)

**Description:**

Improve the performance of the `MapHash` function, mostly by using the
xxhash architecture optimized version.

`hash.Sum` is a 'Go-code' only implementation
`xxhash.Sum64` has optimized versions for different architectures
Both result in the exact same hash though.


For the given benchmarks, the gain is > 10%

From `main`:
```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
BenchmarkMapHashFourItems-16                  	47676003	       236.0 ns/op	      24 B/op	       1 allocs/op
BenchmarkMapHashEightItems-16                 	22551222	       532.3 ns/op	      32 B/op	       2 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap-16    	14098969	       893.1 ns/op	      56 B/op	       3 allocs/op
```

The PR:
```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatautil
cpu: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
BenchmarkMapHashFourItems-16                  	59854737	       203.4 ns/op	      24 B/op	       1 allocs/op
BenchmarkMapHashEightItems-16                 	25609375	       475.0 ns/op	      32 B/op	       2 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap-16    	15950144	       753.8 ns/op	      56 B/op	       3 allocs/op
```

**Testing:**
(Re-)using the same tests and benchmarks to prove semantics didn't
change.

bryan-aguilar committed:
[aws/cwlogs] Reduce noisy logger in CloudWatch logs pusher (#27774)

**Description:** Log at the debug level instead of info level. Existing
behavior would cause excessive log lines on each successful push.

sakulali committed:
[pkg/pdatatest] Ignore trace ID for ptracetest (#27829)

**Description:** 
Support ignore trace ID in span comparisons for ptracetest.

**Link to tracking Issue:**

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27687

sakulali committed:
[pkg/pdatatest] Ignore span ID for ptracetest (#27833)

**Description:** 
Support ignore span ID in span comparisons for ptracetest.

**Link to tracking Issue:** 

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27685

**Testing:** 
make chlog-validate 
go test for pdatatest

**Documentation:**

c-kruse committed:
[processor/k8sattributes] Handle all resource deletion event types (#27847)

**Description:**

The k8s go client's cache expects OnDelete handlers to handle objects of
type DeletedFinalStateUnknown when the cache's watch mechanism misses a
delete and notices later. This changes the processor to handle such
deletes as if they were normal, rather than logging an error and
dropping the change.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27632

**Testing:**

Only what you see in the unit tests. I am open to suggestions, but I
don't see this being a code path we can reasonably cover in the e2e test
suite.

Verified manually locally on a kind cluster.
* Stood up two deployments loosely based off e2e testing resources, one
w/ a collector built from this branch and the other
docker.io/otel/opentelemetry-collector-contrib:latest.
* Both included an additional container in the collector pod I used to
fiddle with iptables rules.
* Added rules to reject traffic to/from the kube api server
* Deleted some namespaces containing deployments generating telemetry.
* Restored connectivity by removing the iptables rules.
* Observed the collector built from this branch was silent (aside from
the junk the k8s client logs due to the broken connection)
* Observed the latest
([0.87.0](https://hub.docker.com/layers/otel/opentelemetry-collector-contrib/0.87.0/images/sha256-77cdd395b828b09cb920c671966f09a87a40611aa6107443146086f2046f4a9a?context=explore))
collector logged a handful of errors for the deleted resources
(api_v1.Pod, and apps_v1.ReplicaSet. I probably just didn't wait long
enough for Namespace.)

```
2023-10-19T02:18:37.781Z        error   kube/client.go:236      object received was not of type api_v1.Pod      {"kind": "processor", "name": "k8sattributes", "pipeline": "metrics", "received": {"Key":"src1/telemetrygen-patched-766d55cbcb-8zktr","Obj":{"metadata":{"name":"telemetrygen-patched-766d55cbcb-8zktr","namespace":"src1","uid":"be5d2268-c8b0-434d-b3b8-8b18083c7a8b","creat
ionTimestamp":"2023-10-19T02:01:08Z","labels":{"app":"telemetrygen-patched","pod-template-hash":"766d55cbcb"},"ownerReferences":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"telemetrygen-patched-766d55cbcb","uid":"a887d67a-d5d6-4269-b520-45dbb4f1cd82","controller":true,"blockOwnerDeletion":true}]},"spec":{"containers":[{"name":"telemetrygen","image":"localhost/telemetrygen
:latest","resources":{}}],"nodeName":"manual-e2e-testing-control-plane"},"status":{"podIP":"10.244.0.56","startTime":"2023-10-19T02:01:08Z","containerStatuses":[{"name":"telemetrygen","state":{},"lastState":{},"ready":false,"restartCount":0,"image":"","imageID":"","containerID":"containerd://2821ef32cd8bf93a13414504c0f8f0c016c84be49d6ffdbd475d7e4681e90c51"}]}}}}
github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor/internal/kube.(*WatchClient).handlePodDelete
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor@v0.87.0/internal/kube/client.go:236
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete
        k8s.io/client-go@v0.28.2/tools/cache/controller.go:253
...

2023-10-19T02:19:03.970Z        error   kube/client.go:868      object received was not of type apps_v1.ReplicaSet      {"kind": "processor", "name": "k8sattributes", "pipeline": "metrics", "received": {"Key":"src1/telemetrygen-stable-5c444bb8b8","Obj":{"metadata":{"name":"telemetrygen-stable-5c444bb8b8","namespace":"src1","uid":"d37707ff-b308-4339-8543-a1caf5705ea8","creationTimestamp":null,"ownerReferences":[{"apiVersion":"apps/v1","kind":"Deployment","name":"telemetrygen-stable","uid":"c421276e-e1bf-40c5-85e1-e92e30363da5","controller":true,"blockOwnerDeletion":true}]},"spec":{"selector":null,"template":{"metadata":{"creationTimestamp":null},"spec":{"containers":null}}},"status":{"replicas":0}}}}
github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor/internal/kube.(*WatchClient).handleReplicaSetDelete
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sattributesprocessor@v0.87.0/internal/kube/client.go:868
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete
        k8s.io/client-go@v0.28.2/tools/cache/controller.go:253
k8s.io/client-go/tools/cache.(*processorListener).run.func1
        k8s.io/client-go@v0.28.2/tools/cache/shared_informer.go:979
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
...
```
**Documentation:** N/A - it is not clear to me whether or not this
should land on the changelog. Its impact on users is marginal.

Signed-off-by: Christian Kruse <ctkruse99@gmail.com>

gord02 committed:
[pkg/ottl] Add Double converter (#27457)

**Description:** Adding a Double converter to pkg/ottl

**Link to tracking Issue:** closes #22056

bagmeg committed:
[chore][exporter/fileexporter] use errors.Join instead of go.uber.org/multierr (#27835)

**Description:** fileexporter: use errors.Join instead of go.uber.org/multierr

**Link to tracking Issue:** #25121

songy23 committed:
[connector/datadog] Allow export to traces pipelines (#27846)

**Description:** <Describe what has changed.>
Allow datadogconnector export to traces pipelines

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were
added.>

**Documentation:** <Describe the documentation added.>

---------

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

TylerHelmuth committed:
[receiver/k8sobjects] Bump to beta stability for logs (#27635)

**Description:**
Bumps the receiver to beta stability for logs

**Link to tracking Issue:** <Issue number if applicable>
Closes
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21503

bagmeg committed:
[chore][exporter/elasticsearchexporter] use errors.Join instead of go.uber.org/multierr (#27857)

**Description:** 
use errors.Join instead of go.uber.org/multierr

**Link to tracking Issue:** <Issue number if applicable>
#25121 

**Testing:** <Describe what testing was performed and which tests were
added.>

**Documentation:** <Describe the documentation added.>

opentelemetrybot committed:
Add JMX metrics gatherer version 1.31.0-alpha (#27852)


cc @open-telemetry/java-contrib-approvers

crobert-1 committed:
[chore][cmd/otelcontribcol] Exclude from linter (#27865)

**Description:**
The linter runs are taking longer and longer as more components and
dependencies are added, it's preferable to simply run on each individual
component.

**Link to tracking Issue:** 
Related: #27850

djaglowski committed:
[chore][pkg/stanza] Log matching paths (#27859)

This PR adds a debug log to give visibility into the exact outcome of
file matching configuration.

It also removes some fragile logging expectations from batching tests. I
believe the meaningful part of the tests remain intact.

VihasMakwana committed:
[extension/encoding/jsonlogencoding] add new extension for JSON codec (#27484)

**Description:** Create a new extension for JSON. This will be used in
pulsarreceiver/kafkareceiver to populate the log record's map from the
raw body.

**Link to tracking Issue:**
[6272](https://github.com/open-telemetry/opentelemetry-collector/issues/6272)

---------

Co-authored-by: Dmitrii Anoshin <anoshindx@gmail.com>

sakulali committed:
[pkg/pdatatest] Ignore span attribute value (#27861)

**Description:**
Support ignore span attribute value in span comparisons for ptracetest.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27689

dmitryax committed:
[cmd/otelcolcontrib] [chore] Send test data in lifecycle processor tests (#27848)

The lifecycle test passes readonly data to processors claimed as not
mutable to catch invalid behavior.

Similar to
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27825,
but for processors tests.

atoulme committed:
[chore] remove mentions of docs.signalfx.com from docs (#27869)

The website docs.signalfx.com has been decommissioned and no longer
resolves. Changing all doc links to the new doc website or removing
them.

djaglowski committed:
[pkg/stanza] Move flush state into reader package (#27843)

martin-majlis-s1 committed:
Extract exportSettings into structure + add more tests

martin-majlis-s1 committed:
Merge branch 'main' into 27651-make-resource-scope-more-flexible

martin-majlis-s1 commented:
I have added more tests to cover also the configuration options for traces.

martin-majlis-s1 commented:
I have messed up the merge so it added whole world as reviewers. 🙈
